test/test_torch.py

```diff
@@ -13052,8 +13067,23 @@ def check_sum_all(tensor):
         else:
             check_sum_all(torch.tensor([True, False, True], dtype=torch.bool, device=device))
 
-    def _test_memory_format_transformations(self, device, input_generator_fn, transformation_fn,
-                                            compare_data=True, default_is_preserve=False):
+    @onlyCPU
+    @dtypes(torch.bfloat16)
+    def test_sum_precision(self, device, dtype):
+        x1 = torch.rand(100, 100)
+        res1 = torch.sum(x1, 1)
+        x2 = x1.bfloat16()
+        res2 = torch.sum(x2, 1)
+        self.assertEqual(res1, res2, 2e-1)
```

Comment:
This test is not testing what you want to test. Proper strategy is
1) have a bfloat16 tensor and float tensor consisting of values exactly representable in bfloat
2) perform a reduction
3) convert float result to number exactly representable in bfloat16
4) verify that bfloat16 and converted results match exactly
This does not require setting absolute tolerance that would depend on the size of the tensor being reduced.

Reply:
This test is used to measure bfloat16 sum precision. If the float tensor is converted from a bfloat16 tensor, there is no change in precision and the test cannot measure bfloat16 sum precision. In most cases, the input/source tensors (like weight, input data, etc.) are float tensors. This test simulates the real case and then measures the bfloat16 sum precision.

Reply:
For this test, we are not interested in theoretical discussions on how good bfloat16 type is and what accuracy it provides compared to fp32. Those questions are answered just by knowing the formats, ULPs and distributions. This test should be testing correct implementation (i.e., that reduction operations on bfloat16 accumulate on fp32). For this, the way to go is an algorithm I provided above.